File: kotlin/sieve/src/main/kotlin/Sieve.kt
object Sieve {

    fun primesUpTo(upperBound: Int): List<Int> {
        val candidates = (2 .. upperBound).toMutableList()
        val result = mutableListOf<Int>()

        while (candidates.isNotEmpty()) {
            val prime = candidates.first()
            result.add(prime)
            candidates.removeIf { it % prime == 0 }
        }

        return result
    }
}

File: swift/sieve/Sources/Sieve/Sieve.swift
private func sieve(candidates: [Int], primes: [Int]) -> [Int] {
  guard let prime = candidates.first else {
    return primes
  }

  return sieve(
    candidates: candidates.filter { !$0.isMultiple(of: prime) },
    primes: primes + [prime]
  )
}

func sieve(limit: Int) -> [Int] {
  guard limit > 1 else {
    return []
  }

  return sieve(candidates: Array(2...limit), primes: [])
}
